Handle unsupported request content type

raphael opened the pull request:
Return 415 Unsupported Media Type when the request decoder does not support the request content type.

raphael committed:
Handle unsupported request content type

Return 415 Unsupported Media Type when the request decoder does not support the request content type.

raphael committed:
Make linter happy

tchssk commented on http/encoding.go:
```diff
@@ -25,6 +27,13 @@ const (
+	// supported.
+	ErrUnsupportedMediaTypeName = "ErrUnsupportedMediaType"
+)
+
```
Should it be defined in `pkg/error.go`?

```go
const (
	UnsupportedMediaType = "unsupported_media_type"
)

func UnsupportedMediaTypeError(ct string) error {
	return PermanentError(UnsupportedMediaType, "unsupported media type %s", ct)
}
```

It seems better to define the error name in snake case like the existing definitions.

* `missing_payload`
* `decode_payload`
* `invalid_field_type`
* `missing_field`
* `invalid_enum_value`
* `invalid_format`
* `invalid_pattern`
* `invalid_range`
* `invalid_length`

raphael replied:
Thank you for the comment! Yes I agree that the name should be snake case for consistency. `pkg/error.go` contains service level errors while this is a HTTP transport level error so it might be better to keep this error definition in the `http` package.

tchssk replied:
`MissingPayloadError()` and `DecodePayloadError()` are only used for HTTP, so aren't they HTTP transport level errors?

raphael replied:
Makes sense, I moved the error definition to `pkg/error.go`.

tchssk commented on http/error.go:
```diff
@@ -81,6 +81,9 @@ func (resp *ErrorResponse) MarshalXML(e *xml.Encoder, _ xml.StartElement) error
 func (resp *ErrorResponse) StatusCode() int {
+	if resp.Name == ErrUnsupportedMediaTypeName {
+		return http.StatusUnsupportedMediaType
+	}
```
It seems that 400 is returned instead of 415 because the error name is overwritten by `DecodePayloadError()`.

https://github.com/goadesign/goa/blob/feature/unsupported_media_type/http/codegen/templates/request_decoder.go.tpl#L25

 If the error is `ServiceError`, it must be returned as is.

```go
var gerr *goa.ServiceError
if errors.As(err, &gerr) {
	return nil, err
}
```

raphael replied:
Great catch! I have modified the generated code to follow the suggestion above.

raphael committed:
Do not wrap decode errors that are already service errors

This makes it possible for the decoding code to return proper HTTP statuses depending on the name of error.
As of this PR, this is only used to return 415 "Unsupported Media Type" when an unsupported content type is used to encode the request payload.

raphael committed:
Refactor slightly to group error definitions

tchssk reviewed (approved):
LGTM!
